README.md
<p align="left">
  <a href="#">
    <img 
      alt="dicom-populate-python-script-logo" 
      src="https://raw.githubusercontent.com/natanaelfneto/dicompopulate/master/assets/dp-logo.png" 
      width="240" />
  </a>
</p>

**DICOM Populate** is small code to send folders of dicom files to as many desired PACS server receivers as possible\
Version: **0.5**
***
# Table of Contents
* [Getting Started](#getting-started)
    * [Installation process](#installation-process)
    * [Dependencies](#dependencies)
    * [Usage](#usage)
    * [TODO](#TODO)
* [License](#license)
***
## Getting Started
### Via Bash
#### Installation Process
* Via pip\
_still not available in Pypi repository_
```Shell
pip install git+https://github.com/natanaelfneto/dicompopulate.git
```
* Via Git\
Clone or download the repository at:
```Shell
git clone https://github.com/natanaelfneto/dicompopulate.git
cd dicompopulate
python setup.py install
```
_enjoy_
***
### Dependencies
- pydicom
- pynetdicom
## TODO
* add _'localhost'_, _'pacs.example.com'_, names support insted or just pure IP Addresses
* fully migrate to python 3 standards
* make C-STORE better by stop closing conections and reopening for each dicom file sent [not yet sure how]
* check if file was save on destination conection (probably with C-FIND)
* check for valid TCP/IP:PORT listen before a C-ECHO attempt
* manage to save (proably with a simple sqlite database or other) wich files were already sent to avoid duplicate attempts on a re-run
* run script for a local and remote origin and local and remote destiny (only availabe so far from local origin to local and remote destiny)
* make available a test script for sending and receiving dicom files on python sender to python receiver, python sender and non-python receiver
## Usage
This messagem can also be found with **python populate.py -h** command
```ShellSession
usage: populate.py [-h] -p PATHS [PATHS ...] -c CONECTIONS [CONECTIONS ...]  [-d] [-v] [-w] [--verbose]

a script to populate a PACS with folder of DICOM files

optional arguments:

-h, --help                                  show this help message and exit

-p PATHS [PATHS ...], 
--paths PATHS [PATHS ...]                   array of dicom folders or files paths

-c CONECTIONS [CONECTIONS ...],
--conections CONECTIONS [CONECTIONS ...]    array of conection parameters for dicom 
                                            receivers (Application Entities)

-d, --debug                                 set debug flag (it only shows debug
                                            information and can be combined with the 
                                            verbose flag for a more robust output and log)

-v, --version                               output software version

-w, --max_workers                           set the maximum number of parallel
                                            processes allowed for C-STORE threads
                                            (default value is 10)

--verbose                                   set verbose flag to enhence output info
                                            (it only shows output information and can
                                            be combined with debug flag for a more
                                            robust output and log)
```
## Examples
### As terminal command
To run it as command:
```Shell
python populate.py --paths /PACS/1/ /PACS/2/ /PACS/3/dicom.dcm --conections DCM4CHEE@10.0.0.1:11112 OTHER@127.0.0.1:5555
```
### As python module
To use it as module follow the exemple, also available in populate/examples/exemple1.py
```Python
from populate import populate

# get all files and paths to send
path_1 = '/PACS/1/'
path_2 = '/PACS/2/'
path_3 = '/PACS/3/dicom.dcm'

# get all desired conections to receive
c_1 = 'DCM4CHEE@10.0.0.1:11112'
c_2 = 'OTHER@127.0.0.1:5555'

# populate
populate.run(
  # debug=True,
  # verbose=True,
  # max_workers=10,
  path=[ path_1, path_2, path_3 ],
  conections=[ c_1, c_2 ]
)

# debug, verbose and max_workers values are optional
```
## License
MIT License

Copyright (c) 2017 Natanael F. Neto (natanaelfneto)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Released]

## [Unreleased]
### 0.5 - 2018-09-15
#### Changed
- populate class call from instance to enable 'with' statement call

#### Added
- process execution timer
- send_helper() function for DRY on sending standalone files and files from folder loops
- multithread handler for the store function
- maximum workers value to be parsed as an optional input value from user. default is set to 10
- workers wait function to release populate instance only when all threads were done and when release flag is set

### 0.4 - 2018-09-12
#### Changed
- logger routines to separated class Logger
- change logger from class input to global variable

#### Added
- verbose flag
- separated verbose and debug outputs but still available for combination of flags
- pydicom module for checking if file is dicom valid
- function to echo validade aet@ip:port conections
- function to store dicom to valid aet@ip:port

### 0.3 - 2018-09-09
#### Changed
- renamed run.py to main.py
- removed python-tail due to hard usage and buggy package name

### Added
- possibility for multi paths to be sent with a PathValidity Class
- possibility for multi pacs creceivers with a ConectionValidity Class
- log folder generation insted of loosed files

### 0.2 - 2018-09-08
#### Changed
- replace of the console output and log file writing with the python logging and a python-tail modules
- renamed dicom_populate.py to run.py

#### Added
- modules folder for non pip third party modules
- assets folder for better organization

### 0.1 - 2018-09-07
#### Added
- dcm4chee-2.0.29 binaries
- dicom_populate.py script with basic run of DCMSND for dicom files inside folder and subfolders
- a very basic usage og a log file and console output
- argsparser module

### 0.0 - 2018-09-07
#### Added
- project created

dicom_populate.py
from __future__ import print_function
import  argparse, datetime, os, time

# get current datetime
st = datetime.datetime.fromtimestamp(time.time()).strftime("%Y-%m-%d %H:%M:%S")

# parse arguments from terminal
parser = argparse.ArgumentParser(description='a DCM4CHEE binaries usage with python scripting')
parser.add_argument('-p','--path', help='dicom folder path', required=True, default="check_string_for_empty")
parser.add_argument('-c','--pacs', help='ae title of destination PACS', required=True, default="check_string_for_empty")
parser.add_argument('-a','--address', help='ip address of destination PACS', required=True, default="check_string_for_empty")
parser.add_argument('-P','--port', help='port of destination PACS', required=True, default="check_string_for_empty")
parser.add_argument('-v','--verbose', action='store_true', help='process verbose flag', required=False)
parser.add_argument('-l','--log', action='store_true', help='save log file', required=False)
# add arguments on array
args = parser.parse_args()
log_instance = 0
log = "populate_"+str(log_instance)+".log"
dmc_log = "dcm.log"

# check log flag
if args.log:
    # check if log file already exist
    while os.path.isfile(os.getcwd()+"/"+log):
        log_instance = log_instance + 1
        log = "populate_"+str(log_instance)+".log"
    if not os.path.isfile(os.getcwd()+"/"+log):
        f = open(log,'w')
        f.write(st+" Starting log file\n")

ssh = args.pacs+"@"+args.address+":"+args.port
output = st+" Sending dicom files located at "+args.path+" to "+ssh

# check verbose flag
if args.verbose:
    print(output)
if args.log:
    f.write("\n"+output+"\n")

files = []
i = 0

# loop through folder
for root, dirs, files in os.walk(args.path):
    # check if folder is not empty
    if files:
        for file in files:
            # get absolute path
            file_path = os.path.abspath(os.path.join(root, file))
            # set command for dcm4chee binary
            command = "nohup sh ./dcm4chee-2.0.29/bin/dcmsnd "+ssh+" "+file_path+" </dev/null> "+dmc_log+" 2>&1 &"
            i = i + 1
            output = st+" Sending "+str(i)+": "+file_path
            if args.log:
                f.write(output+"\n")
            if args.verbose:
                print(output, end='\n')
            else:
                print(output, end='\r')
            # run the command
            os.system(command)
print("",end='\n')
if args.verbose:
    print("Log file merged. \n"+st+" "+"Process is now finished. \n The log file can be found at "+os.getcwd()+outlog)
# finish

main.py
#!/usr/bin/env python
from __future__ import print_function
# third party imports
import argparse
import logging
import os
import re
import sys
# local imports

# module version
__version__ = "0.3"

# main class
class Populate(object):
    # represents a tail command
    def __init__(self):
        ''' 
            Initiate a DICOM Populate instance.
            
            Arguments for send():
                files: Array of files and/or folder to be sent to a PACS environment
                aet: Application Entity Title, the PACS 'given name'
                addr: short for address, the IP Address of the server wich is runnig
                port: usually 11112 for dicom comunication, but customable
        '''
    # 
    def send(self, paths, conections):
        # set basic variables
        files = []  # array of files
        i = 0       # file counter
        # loop through folder
        logger.debug('Looping throug folder and subfolders...')
        for path in paths:
            for root, dirs, files in os.walk(path):
                # check if folder is not empty
                if files:
                    for file in files:
                        # get absolute path
                        file_path = os.path.abspath(os.path.join(root, file))
                        # increment file counter
                        i = i + 1
                        for conection in conections:
                            logger.info(
                                'Sending file %s: %s to %s@%s:%s',
                                str(i),
                                file_path,
                                conection['ae_title'],
                                conection['addr'],
                                conection['port']
                                )
                else:
                    # if no file is found inside folder, log it down
                    root = os.path.abspath(os.path.join(root))
                    logger.debug('No file found inside this folder %s', root)
            logger.info('Finished sending all files from %s', path)
        logger.info('Finished sending all files')

# paths argument parser
class PathsValidity(object):
    # path validity init
    def __init__(self):
        ''' 
            Initiate a DICOM Populate Path Validity instance.
        '''
    # path validity checker function
    def checker(self, paths):
        logger.debug('checking validity of parsed files and folders')
        for path in paths:
            if not os.access(path, os.F_OK) and not os.access(path, os.R_OK):
                logger.debug( \
                    "Path '%s' could not be found or does not have read permitions, \
                    therefore will be ignored", path
                    )
                paths.remove(path)
        logger.debug("Paths '%s' was/were found", paths)

        return paths

# conection argument parser
class ConectionValidity(object):
    # path validity init
    def __init__(self):
        ''' 
            Initiate a DICOM Populate Conection Validity instance.
        '''
    # conection validity checker function
    def checker(self, conections):
        logger.debug('checking validity of parsed conections')
        # set basic variables
        aes = []    # array for AEs
        for conection in conections:
            # check if conection format passed is correct
            if not conection.count('@') == 1 or not conection.count(':') == 1:
                logger.debug('Wrong conection format was passed: %s', conection)
            else:
                # get AE Title format
                ae_title = conection.split('@')[0]
                # get TCP/IP Address format
                addr = conection.split('@')[1].split(':')[0]
                # get TCP/IP Port format
                port = conection.split('@')[1].split(':')[1]
                # check if ae_title, address and port are in correct format
                if not re.match(r"^\w+$",ae_title):
                    logger.debug('Wrong conection AE Title was passed: %s', ae_title)
                elif not re.match(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$",addr):
                    logger.debug('Wrong conection TCP/IP Address was passed: %s', addr)
                elif not re.match(r"^\d{4,5}$",port):
                    logger.debug('Wrong conection TCP/IP Port was passed: %s', port)
                else:
                    aes.append({
                        'ae_title': ae_title,
                        'addr': addr,
                        'port': port
                    })

        return aes

if __name__ == "__main__":
    # argparser init
    parser = argparse.ArgumentParser(
        description='a sender for folders of dicom files in python'
    )
    # path argument parser
    parser.add_argument(
        '-p','--paths',
        nargs='+',
        help='dicom folders or files path', 
        default="check_string_for_empty",
        required=True
    )
    # conection argument parser
    parser.add_argument(
        '-c','--conections',
        nargs='+',
        help='the conection parameters for dicom receivers',
        default="check_string_for_empty",
        required=True
    )
    # debug flag argument parser
    parser.add_argument(
        '-d','--debug',
        action='store_true', 
        help='process debug flag',
        default=False,
        required=False
    )
    # verbose flag argument parser
    parser.add_argument(
        '-v','--verbose',
        action='store_true', 
        help='runtime verbose flag',
        default=False,
        required=False
    )
    # passing filtered arguments as array
    args = parser.parse_args() 

    # setting logging basic config avriables
    log_folder = 'log'
    if args.debug:
        log_level = 'DEBUG'
        log_file = log_folder+'/populate_debug.log'
        log_format = '%(asctime)s %(name)s %(levelname)s %(message)s'
        log_date_format = '%Y-%m-%d %H:%M:%S'
    else:
        log_level = 'INFO'
        log_file = log_folder+'/populate_info.log'
        log_format = '%(asctime)s %(levelname)s %(message)s'
        log_date_format = '%Y-%m-%d %H:%M:%S'
    
    if not os.path.exists(log_folder):
        os.makedirs(log_folder)

    # setting logging basic config
    logging.basicConfig(
        level=getattr(logging,log_level),
        format=log_format,
        datefmt=log_date_format,
        filename=log_file,
        filemode='a+'
        )
    logger = logging.getLogger(__name__)

    # temporary tail with terminal execution
    if args.verbose:
        os.system('tail -f populate_info.log &')
    if args.verbose and args.debug:
        os.system('tail -f -n 1 populate_*.log &')

    # check validity of the paths parsed
    paths = PathsValidity()
    paths = paths.checker(args.paths)

    # check validity of the paths parsed
    conections = ConectionValidity()
    conections = conections.checker(args.conections)

    populate = Populate()
    populate.send(paths, conections)